rfcs/0001-messages.md

```diff
@@ -0,0 +1,301 @@
+# FROST messages
+
+Proposes a message layout to exchange information between participants of a FROST setup using the [jubjub](https://github.com/zkcrypto/jubjub) curve.
+
+## Motivation
+
+Currently FROST library is complete for 2 round signatures with a dealer/aggregator setup.
+This proposal is only considering that specific features, additions and upgrades will need to be made when DKG is implemented.
+
+Assuming all participants have a FROST library available we need to define message structures in a way that data can be exchanged between participants. The proposal is a collection of data types so each side can do all the actions needed for a real life situation.
+
+## Definitions
+
+- `dealer`
+- `aggregator`
+- `signer`
+- `nonce`
+- `commitment`
+- 
+
+## Guide-level explanation
+
+We propose a message separated in 2 parts, a header and a payload:
+
+```rust
+struct Message {
+    header: Header,
+    payload: Payload,
+}
+```
+
+`Header` will look as follows:
+
+```rust
+struct Header {
+    msg_type: MsgType,
+    version: MsgVersion,
+    sender: ParticipantID,
+    receiver: ParticipantID,
+}
+```
+
+While `Payload` will be defined as:
+
+```rust
+enum Payload {
+    DealerBroadcast(MsgDealerBroadcast),
+    Commitments(MsgCommitments),
+    SigningPackage(MsgSigningPackage),
+    SignatureShare(MsgSignatureShare),
+    FinalSignature(MsgFinalSignature),
+}
+```
+
+All the messages and new types will be defined in a new file `src/frost/messages.rs`
+
+## Reference-level explanation
+
+Here we explore in detail the header types and all the message payloads. 
+
+### Header
+
+Fields of the header define new types. Proposed implementation for them is as follows:
+
+```rust
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    DealerBroadcast,
+    Commitments,
+    SigningPackage,
+    SignatureShare,
+    FinalSignature,
+}
+
+struct MsgVersion(u8);
+
+struct ParticipantID(u8);
+```
+
+### Payloads
+
+Each payload defines a new message:
+
+```rust
+/// Dealer must send this message with initial data to each participant involved.
+/// With this, the participant should be able to build a `SharePackage` and use
+///  the `sign()` function.
+/// `public_key` can be calculated from the `secret_key`.
+struct MsgDealerBroadcast {
+    /// The secret key as a frost::Scalar.
+    secret_key: frost::Scalar,
+    /// Commitment for the signer as a single jubjub::AffinePoint.
+    commitment: jubjub::AffinePoint,
+    /// The public signing key that represents the entire group.
+    group_public: GroupPublic,
+}
+
+/// The point and verification bytes needed to generate the group public key
+struct GroupPublic {
+    /// The point
+    point: jubjub::AffinePoint,
+    /// The verification bytes
+    bytes: [u8; 32],
+}
+
+/// Each signer participant send to the aggregator the 2 points
+///  needed for commitment building.
+struct MsgCommitments {
+    /// The commitment the signer is sending.
+    commitment: Commitment,
```

Comment:
These commitments correspond to the FROST preprocessing protocol described in Figure 2 in the paper though. These are commitments to the single-use nonces called `d` and `e` in the paper and `hiding` and `binding` in the implementation. These are different from those sent by the trusted dealer --- the dealer sends commitments to polynomial coefficients. There are multiple types of "commitments" in the paper (and therefore in the implementation as well), so it might be confusing.

Reply:
The FROST preprocessing protocol (presented in Figure 2) puts the commitments into a list, that's why I suggested a vector. If the participants send a vector of commitments, the signer can cache them. This would allow single-round signing, while sending only one pair of commitments (hiding and binding) would enforce two-round signing. I'm not sure which one we actually intend to support now.

Reply:
Preprocessing involves stored state, so it will complicate the implementation.
Let's focus on the simple two round signing protocol for now?

We can add support for single-round signing and cached state later. (We'll need a new protocol version and message type, but that's ok.)

There might also be some security or reliability implications of cached state, which we should work through in a separate design. (How do we make sure the cached state is used for the right signatures? What if participants get out of sync? How do we restart a round when we run out of cached state? How can we test all this?)

Reply:
Great questions. I think we can focus on the two round signing for now. 

I'd say the users of FROST can handle the state by themselves since they have to store their IDs and keys already. They also have to keep track of which signing group the IDs and keys are associated with since one user can participate in multiple groups. (This can be trivially done since each group is uniquely identified by its public key.)

Here's my attempt to answer some of the questions:
> - How do we make sure the cached state is used for the right signatures?

Notice that the aggregator includes the commitments into each `SigningPackage` so that the participants can check the commitments.

> - What if participants get out of sync?
> - How do we restart a round when we run out of cached state?

I think this the users can handle this by themselves too.